src/code/providers/lara-provider.coffee

```diff
@@ -163,32 +164,154 @@ class LaraProvider extends ProviderInterface
+
+        newRunState = _.cloneDeep existingRunState
+        newRunState.docStore = newDocStore
+
+        rawData = JSON.stringify(newRunState)
+        learnerUrl = if newRunState.learner_url? and typeof newRunState.learner_url is "string" then newRunState.learner_url else null
+        learnerParam = if learnerUrl then "&learner_url=#{encodeURIComponent(learnerUrl)}" else ""
+
+        updateRunState = (url, done) ->
+          $.ajax({
+            type: 'PUT'
+            url: "#{url}?raw_data=#{encodeURIComponent(rawData)}#{learnerParam}"
+            dataType: 'json'
+            xhrFields:
+              withCredentials: true
+          })
+          .done (data, status, jqXHR) ->
+            if data?.success is false
+              done("Could not open the specified document because an error occurred [updateState] (#{data.message})")
+            else
+              done(null)
+          .fail (jqXHR, status, error) ->
+            done("Could not open the specified document because an error occurred [updateState]")
+
+        urlQueue = urls.slice()
+        processQueue = ->
+          if urlQueue.length is 0
+            callback null
+          else
+            url = urlQueue.shift()
+            updateRunState url, (err) ->
+              if err
+                callback err
+              else
+                processQueue()
+        processQueue()
+
+      processCreateResponse = (createResponse) =>
+        docStore =
+          recordid: createResponse.id
+          accessKeys:
+            readOnly: createResponse.readAccessKey
+            readWrite: createResponse.readWriteAccessKey
+
+        codapUrl = if window.location.origin \
+                    then "#{window.location.origin}#{window.location.pathname}" \
+                    else "#{window.location.protocol}//#{window.location.host}#{window.location.pathname}"
+        reportUrlLaraParams =
+          recordid: createResponse.id
+          accessKeys:
+            readOnly: createResponse.readAccessKey
+        encodedLaraParams = @encodeParams reportUrlLaraParams
+        existingRunState.lara_options ?= {}
+        existingRunState.lara_options.reporting_url = "#{codapUrl}?launchFromLara=#{encodedLaraParams}"
+
       # Check if we have a document associated with this run state already (2a) or not (2b)
-      rawData = @extractRawDataFromRunState runState
-      docStoreParams = rawData.docStore
-      if docStoreParams? and docStoreParams.recordid? and
-          (docStoreParams.accessKeys?.readOnly? or docStoreParams.accessKeys?.readWrite?)
-        # (2a) load the document associated with this run state
-        metadata.providerData = _.cloneDeep docStoreParams
-        @load metadata, (err, content) =>
-          @client.removeQueryParams @removableQueryParams
-          callback err, content, metadata
-        return
+      if docStore?.recordid? and (docStore.accessKeys?.readOnly? or docStore.accessKeys?.readWrite?)
+
+        cloneDoc = (callback) =>
```

Comment:
Consider moving these embedded functions into their own file for better readability.